File: python_modules/jupedsim/jupedsim/native/simulation.py
# Copyright © 2012-2023 Forschungszentrum Jülich GmbH
# SPDX-License-Identifier: LGPL-3.0-or-later

import py_jupedsim as py_jps

from jupedsim.native.geometry import Geometry
from jupedsim.native.journey import JourneyDescription
from jupedsim.native.models import (
    GCFMModelAgentParameters,
    VelocityModelAgentParameters,
)
from jupedsim.native.stages import (
    ExitProxy,
    NotifiableQueueProxy,
    WaitingSetProxy,
    WaypointProxy,
)
from jupedsim.native.tracing import Trace


class Simulation:
    def __init__(self, model, geometry: Geometry, dt: float) -> None:
        self._sim = py_jps.Simulation(
            model=model, geometry=geometry._obj, dt=dt
        )

    def add_waypoint_stage(
        self, position: tuple[float, float], distance
    ) -> int:
        return self._sim.add_waypoint_stage(position, distance)

    def add_queue_stage(self, positions: list[tuple[float, float]]) -> int:
        return self._sim.add_queue_stage(positions)

    def add_waiting_set_stage(
        self, positions: list[tuple[float, float]]
    ) -> int:
        return self._sim.add_waiting_set_stage(positions)

    def add_exit_stage(self, polygon: list[tuple[float, float]]) -> int:
        return self._sim.add_exit_stage(polygon)

    def add_journey(self, journey: JourneyDescription) -> int:
        return self._sim.add_journey(journey._obj)

    def add_agent(
        self,
        parameters: GCFMModelAgentParameters | VelocityModelAgentParameters,
    ) -> int:
        return self._sim.add_agent(parameters._obj)

    def remove_agent(self, agent_id: int) -> bool:
        return self._sim.remove_agent(agent_id)

    def removed_agents(self) -> list[int]:
        return self._sim.removed_agents()

    def iterate(self, count: int = 1) -> None:
        self._sim.iterate(count)

    def switch_agent_profile(self, agent_id: int, profile_id: int) -> None:
        self._sim.switch_agent_profile(
            agent_id=agent_id, profile_id=profile_id
        )

    def switch_agent_journey(
        self, agent_id: int, journey_id: int, stage_index: int
    ) -> None:
        self._sim.switch_agent_journey(
            agent_id=agent_id, journey_id=journey_id, stage_index=stage_index
        )

    def agent_count(self) -> int:
        return self._sim.agent_count()

    def elapsed_time(self) -> float:
        return self._sim.elapsed_time()

    def delta_time(self) -> float:
        return self._sim.delta_time()

    def iteration_count(self) -> int:
        return self._sim.iteration_count()

    def agents(self):
        return self._sim.agents()

    def agents_in_range(self, pos: tuple[float, float], distance: float):
        return self._sim.agents_in_range(pos, distance)

    def agents_in_polygon(self, poly: list[tuple[float, float]]):
        return self._sim.agents_in_polygon(poly)

    def get_stage_proxy(self, stage_id: int):
        stage = self._sim.get_stage_proxy(stage_id)
        match stage:
            case py_jps.WaypointProxy():
                return WaypointProxy(stage)
            case py_jps.ExitProxy():
                return ExitProxy(stage)
            case py_jps.NotifiableQueueProxy():
                return NotifiableQueueProxy(stage)
            case py_jps.WaitingSetProxy():
                return WaitingSetProxy(stage)
            case _:
                raise Exception(
                    f"Internal error, unexpected type: {type(stage)}"
                )

    def set_tracing(self, status: bool) -> None:
        self._sim.set_tracing(status)

    def get_last_trace(self) -> Trace:
        return self._sim.get_last_trace()

File: python_bindings_jupedsim/simulation.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "Simulation.hpp"

#include "GenericAgent.hpp"
#include "Geometry/Geometry2D.hpp"
#include "Journey.hpp"
#include "OperationalModel.hpp"
#include "Polygon.hpp"
#include "Stage.hpp"
#include "StageDescription.hpp"
#include "conversion.hpp"
#include "type_casters.hpp" // IWYU pragma: keep

#include <pybind11/attr.h>
#include <pybind11/cast.h>
#include <pybind11/detail/common.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h> // IWYU pragma: keep

#include <cstdint>
#include <map>
#include <memory>
#include <stdexcept>
#include <tuple>
#include <vector>

namespace py = pybind11;

void init_simulation(py::module_& m)
{
    py::class_<Simulation>(m, "Simulation")
        .def(
            // The model is moved out of the Python object into Simulation. After this constructor
            // returns, the Python model object passed here is disowned/invalid and must not be
            // reused.
            py::init(
                [](std::unique_ptr<OperationalModel> model, Geometry2D geometry, double dT) {
                    if(!model) {
                        throw std::invalid_argument("model must not be None");
                    }
                    return std::make_unique<Simulation>(
                        std::move(model), std::make_unique<Geometry2D>(geometry), dT);
                }),
            py::kw_only(),
            py::arg("model"),
            py::arg("geometry"),
            py::arg("dt"))
        .def(
            "add_waypoint_stage",
            [](Simulation& sim, std::tuple<double, double> position, double distance) {
                return sim.AddStage(WaypointDescription{intoPoint(position), distance}).getID();
            })
        .def(
            "add_queue_stage",
            [](Simulation& sim, const std::vector<std::tuple<double, double>>& positions) {
                return sim.AddStage(NotifiableQueueDescription{intoPoints(positions)}).getID();
            })
        .def(
            "add_waiting_set_stage",
            [](Simulation& sim, const std::vector<std::tuple<double, double>>& positions) {
                return sim.AddStage(NotifiableWaitingSetDescription{intoPoints(positions)}).getID();
            })
        .def(
            "add_exit_stage",
            [](Simulation& sim, const std::vector<std::tuple<double, double>>& polygon) {
                return sim.AddStage(ExitDescription{Polygon{intoPoints(polygon)}}).getID();
            })
        .def(
            "add_direct_steering_stage",
            [](Simulation& sim) { return sim.AddStage(DirectSteeringDescription{}).getID(); })
        .def(
            "add_journey",
            [](Simulation& sim, std::map<uint64_t, TransitionDescription>& journey) {
                auto native_journey = std::map<BaseStage::ID, TransitionDescription>{};
                for(const auto& [stage_id, desc] : journey) {
                    native_journey.emplace(stage_id, desc);
                }

                return sim.AddJourney(native_journey).getID();
            })
        .def(
            "add_agent",
            [](Simulation& sim,
               uint64_t journeyId,
               uint64_t stageId,
               GenericAgent::ModelState state) {
                return sim
                    .AddAgent(GenericAgent(
                        GenericAgent::ID::Invalid, journeyId, stageId, std::move(state)))
                    .getID();
            },
            py::kw_only(),
            py::arg("journey_id"),
            py::arg("stage_id"),
            py::arg("state"))
        .def(
            "mark_agent_for_removal",
            [](Simulation& sim, uint64_t id) { sim.MarkAgentForRemoval(id); })
        .def(
            "removed_agents",
            [](const Simulation& sim) {
                auto removed_agent_ids = sim.RemovedAgents();
                auto agent_ids = std::vector<GenericAgent::ID::underlying_type>();
                agent_ids.reserve(removed_agent_ids.size());
                for(auto agent_id : removed_agent_ids) {
                    agent_ids.emplace_back(agent_id.getID());
                }
                return agent_ids;
            })
        .def("iterate", [](Simulation& sim) { sim.Iterate(); })
        .def(
            "switch_agent_journey",
            [](Simulation& sim, uint64_t agentId, uint64_t journeyId, uint64_t stageId) {
                sim.SwitchAgentJourney(agentId, journeyId, stageId);
            },
            py::kw_only(),
            py::arg("agent_id"),
            py::arg("journey_id"),
            py::arg("stage_id"))
        .def("agent_count", [](const Simulation& sim) { return sim.AgentCount(); })
        .def("elapsed_time", [](const Simulation& sim) { return sim.ElapsedTime(); })
        .def("delta_time", [](const Simulation& sim) { return sim.DT(); })
        .def("iteration_count", [](const Simulation& sim) { return sim.Iteration(); })
        .def(
            "agents",
            [](Simulation& sim) { return py::make_iterator(sim.Agents()); },
            py::keep_alive<0, 1>())
        .def(
            // TRANSIENT ONLY: the returned object wraps a raw reference into the
            // simulation's agent storage. It must not be stored across iterate();
            // callers (Python agent handles) resolve it freshly inside every
            // property access.
            "agent",
            [](Simulation& sim, uint64_t agentId) -> auto& { return sim.Agent(agentId); },
            py::arg("agent_id"),
            py::return_value_policy::reference)
        .def(
            "agents_in_range",
            [](Simulation& sim, std::tuple<double, double> pos, double distance) {
                auto agents_in_range = sim.AgentsInRange(intoPoint(pos), distance);
                auto agents = std::vector<uint64_t>();
                agents.reserve(agents_in_range.size());
                for(auto agent : agents_in_range) {
                    agents.emplace_back(agent.getID());
                }
                return agents;
            })
        .def(
            "agents_in_polygon",
            [](Simulation& sim, const std::vector<std::tuple<double, double>>& poly) {
                auto agents_in_range = sim.AgentsInPolygon(intoPoints(poly));
                auto agents = std::vector<uint64_t>();
                agents.reserve(agents_in_range.size());
                for(auto agent : agents_in_range) {
                    agents.emplace_back(agent.getID());
                }
                return agents;
            })
        .def("get_stage_proxy", [](Simulation& sim, uint64_t id) { return sim.Stage(id); })
        .def("set_tracing", [](Simulation& sim, bool status) { sim.SetTracing(status); })
        .def(
            "set_timer_log_level",
            [](Simulation& sim, size_t level) { sim.SetTimerLogLevel(level); })
        .def("get_geometry", [](Simulation& sim) { return sim.Geo(); })
        .def(
            "push_timer",
            [](Simulation& sim, const std::string& name, size_t probe_log_level) {
                sim.PushTimer(name, probe_log_level);
            })
        .def("pop_timer", [](Simulation& sim, const std::string& name) { sim.PopTimer(name); })
        .def(
            "get_duration",
            [](Simulation& sim, const std::string_view name) { return sim.GetTimerDuration(name); })
        .def("get_durations", [](Simulation& sim) { return sim.GetTimerDurations(); });
}
